File: Array/NextPermutation.cpp
#include<bits/stdc++.h>

using namespace std;

void nextPermutation(vector<int> &nums) {
    int sizeOfArray = nums.size();

    int firstDecreasingFromLeft = -1;
    int i = sizeOfArray - 1;

    while(i > 0) {
        if(nums[i] > nums[i -1]) {
            firstDecreasingFromLeft = i - 1;
            break;
        }
        i--;
    }

    if(firstDecreasingFromLeft == -1) {
        reverse(nums.begin(), nums.end());
        return;
    }

    i = sizeOfArray - 1;

    while(i < firstDecreasingFromLeft) {
        if(nums[i] > nums[i - 1])
            break;

        i--;
    }

    // i holds the position of first element from the right
    // that is greater than its left

    swap(nums[i], nums[firstDecreasingFromLeft]);
    i = sizeOfArray - 1;

    for (int j = firstDecreasingFromLeft + 1; j < i; j++, i--)
        swap(nums[j], nums[i]);
}

File: Array/NextPermutation.java
package Array;

public class NextPermutation {
    /*
     * https://www.geeksforgeeks.org/dsa/next-permutation/
     */
    
    void nextPermutation(int[] arr) {

        int n = arr.length;

        int firstElSmallerThanItsRight = -1;

        int i = n - 2;

        // The first number to be moved is the rightmost number smaller than its next.
        while (i >= 0) {
            if (arr[i] < arr[i + 1]) {
                firstElSmallerThanItsRight = i;
                break;
            }
            i--;
        }

        if (firstElSmallerThanItsRight == -1) {
            reverese(arr, 0, n - 1);
            return;
        }

        i = n - 1;

        // The number to come in-place is the first number from right, greater than pivot(firstElSmallerThanItsRight).
        while (i > firstElSmallerThanItsRight) {
            if (arr[i] > arr[firstElSmallerThanItsRight]) {
                // if found, swap them and sort the right side
                swap(arr, firstElSmallerThanItsRight, i);
                break;
            }
            i--;
        }

        reverese(arr, firstElSmallerThanItsRight + 1, n - 1);

    }
    
    private void reverese(int arr[], int left, int right) {
        while (left < right) {
            swap(arr, left, right);
            left++;
            right--;
        }
    }

    private void swap(int arr[], int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }
    
}
